File: day01/day01.clj
(ns day01
  (:require [clojure.string :refer [split-lines]]))

(defn parse-file [s]
  (map read-string (split-lines s)))

(defn sliding-sum [l]
  (map (partial reduce +) (partition 3 1 l)))

(defn larger [l]
  (count (filter (partial apply <) (partition 2 1 l))))

(defn -main []
  (->> (slurp "input")
       parse-file
       sliding-sum
       larger))

File: day01/day01.hs
module Main where

parseFile :: String -> [Int]
parseFile = map read . lines

sum3 :: Int -> Int -> Int -> Int
sum3 a b c = a + b + c

slidingSum :: [Int] -> [Int]
slidingSum (x:y:xs) = zipWith3 sum3 (x:y:xs) (x:xs) xs

larger :: [Int] -> Int
larger l = length . filter (uncurry (<)) . zip l $ tail l

main :: IO ()
main = do
  nums <- slidingSum . parseFile <$> readFile "input"
  print $ larger nums
